File: src/code/java/data_structures/binary_tree.java
public class TreeNode<T> {
    T data;
    TreeNode<T> left;
    TreeNode<T> right;

    public TreeNode(T data) {
        this.data = data;
        left = null;
        right = null;
    }
}

public class BinaryTree<T> {
    TreeNode<T> root;

    public BinaryTree() {
        root = null;
    }

    public void insert(T data) {
        root = insertNode(root, data);
    }

    private TreeNode<T> insertNode(TreeNode<T> node, T data) {
        if (node == null) {
            return new TreeNode<>(data);
        }

        if (node.left == null) {
            node.left = new TreeNode<>(data);
        } else if (node.right == null) {
            node.right = new TreeNode<>(data);
        } else {
            node.left = insertNode(node.left, data);
        }

        return node;
    }

    @Override
    public String toString() {
        return root == null ? "Empty tree" : printTree(root, "", true);
    }

    private String printTree(TreeNode<T> node, String prefix, boolean isLeft) {
        if (node == null) {
            return "";
        }

        StringBuilder sb = new StringBuilder();
        sb.append(printTree(node.right, prefix + (isLeft ? "│   " : "    "), false));
        sb.append(prefix).append(isLeft ? "└── " : "┌── ").append(node.data).append("\n");
        sb.append(printTree(node.left, prefix + (isLeft ? "    " : "│   "), true));

        return sb.toString();
    }
}

File: src/code/ruby/data_structures/binary_tree.rb
class TreeNode
  attr_accessor :data, :left, :right

  def initialize(data)
    @data   = data
    @left   = nil
    @right  = nil
  end
end

class BinaryTree
  attr_accessor :root

  def initialize
    @root = nil
  end

  def insert(data)
    if @root.nil?
      @root = TreeNode.new(data)
    else
      insert_node(@root, data)
    end
  end

  def insert_node(node, data)
    return TreeNode.new(data) if node.nil?

    if node.left.nil?
      node.left = TreeNode.new(data)
    elsif node.right.nil?
      node.right = TreeNode.new(data)
    else
      node.left = insert_node(node.left, data)
    end

    node
  end

  def to_s
    @root.nil? ? 'Empty tree' : print_tree(@root, '', true)
  end

  private

  def print_tree(node, prefix, is_left)
    return '' if node.nil?

    result = ''
    result += print_tree(node.right, prefix + (is_left ? '│   ' : '    '), false)
    result += prefix + (is_left ? '└── ' : '┌── ') + node.data.to_s + "\n"
    result += print_tree(node.left, prefix + (is_left ? '    ' : '│   '), true)

    result
  end
end
